aptos-move/aptos-vm/src/move_vm_ext/warm_vm_cache.rs

```diff
@@ -0,0 +1,162 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+#![forbid(unsafe_code)]
+
+use crate::{counters::TIMER, move_vm_ext::AptosMoveResolver};
+use aptos_framework::natives::code::PackageRegistry;
+use aptos_infallible::RwLock;
+use aptos_metrics_core::TimerHelper;
+use aptos_types::on_chain_config::OnChainConfig;
+use bytes::Bytes;
+use move_binary_format::errors::{Location, PartialVMError, VMResult};
+use move_core_types::{
+    account_address::AccountAddress,
+    ident_str,
+    identifier::Identifier,
+    language_storage::{ModuleId, CORE_CODE_ADDRESS},
+    vm_status::StatusCode,
+};
+use move_vm_runtime::{config::VMConfig, move_vm::MoveVM, native_functions::NativeFunction};
+use once_cell::sync::Lazy;
+use std::collections::HashMap;
+
+const WARM_VM_CACHE_SIZE: usize = 8;
+
+pub(crate) struct WarmVmCache {
+    cache: RwLock<HashMap<WarmVmId, MoveVM>>,
+}
+
+static WARM_VM_CACHE: Lazy<WarmVmCache> = Lazy::new(|| WarmVmCache {
+    cache: RwLock::new(HashMap::new()),
+});
+
+impl WarmVmCache {
+    pub(crate) fn get_warm_vm(
+        natives: impl IntoIterator<Item = (AccountAddress, Identifier, Identifier, NativeFunction)>,
+        vm_config: VMConfig,
+        resolver: &impl AptosMoveResolver,
+    ) -> VMResult<MoveVM> {
+        WARM_VM_CACHE.get(natives, vm_config, resolver)
+    }
+
+    fn get(
+        &self,
+        natives: impl IntoIterator<Item = (AccountAddress, Identifier, Identifier, NativeFunction)>,
+        vm_config: VMConfig,
+        resolver: &impl AptosMoveResolver,
+    ) -> VMResult<MoveVM> {
+        let _timer = TIMER.timer_with(&["warm_vm_get"]);
+        let id = {
+            let _timer = TIMER.timer_with(&["get_warm_vm_id"]);
+            WarmVmId::new(&vm_config, resolver)?
+        };
+
+        if let Some(vm) = self.cache.read().get(&id) {
+            let _timer = TIMER.timer_with(&["warm_vm_cache_hit"]);
+            return Ok(vm.clone());
+        }
+
+        {
+            let _timer = TIMER.timer_with(&["warm_vm_cache_miss"]);
+            let mut cache_locked = self.cache.write();
+            if let Some(vm) = cache_locked.get(&id) {
+                // Another thread has loaded it
+                return Ok(vm.clone());
+            }
+
+            let vm = MoveVM::new_with_config(natives, vm_config)?;
+            Self::warm_vm_up(&vm, resolver);
+
+            // Not using LruCache because its `::get()` requires &mut self
```

Comment:
Hm. don't we expect the use pattern to just be going forward, mainly? I.e. wouldn't we get most of the benefits if we just kept recently added VMs (instead of not recently getted)? If so, can't we use a simpler data-structure? (even RwLock<Vec<>> if we wanted to keep a few around)

Reply:
`RwLock<VecDeque>` sounds reasonable, but I think it's straightforward and easy to read with HashMap..